### DISCUSSION TITLE
Protected extension methods

### DISCUSSION THREAD

**Discussion by Audionysos** [dotnet/csharplang#2081] [Language Ideas]
**The idea:**

- A programmer can access protected member of `this` object inside extension method.
- When a protected member is used within extension method
  - The method cannot use any private or protected members of the type where it is defined.
  - The method is only visible in scope of class that directly derive from type of `this` and that also have access to all of the members used inside extension method. *(alternatively access from extension method could be restricted only to parent type public members and members of type of `this`).*
  - The method can only be called with `this` argument object which is the same object that is accessible using `this` keyword in caller method (i.e. only current instance of an object can be used by the extension method)
- The method could be just copy-paste as private method into class that is using it at compile time.

**Usage**:

```c#
// Not accessible library.
public class Car {
  protected void makeASound(Sound s) {
    //...
  }
}

// My library
public class ModelA : Car {

  private void onRun() {
    MakeEngineSound();
    //...
  }

  private void onBreaking() {
    MakeSqueakinSound();
  }
}

public static class MyCarSoundsExtensions {
  public static void MakeEngineSound(this Car c) {
    c.makeASound(new Sound(/*...*/));
  }

}

// My friend's library
public static class MyFriensCarSoundsExtensions {
  public static void MakeSqueakinSound(this Car c) {
    c.makeASound(new Sound(/*...*/));
  }

}
```

And the compiler will just insert the extension methods inside `ModelA` class:

```c#
public class ModelA : Car {

  private void onRun() {
    MakeEngineSound();
    //...
  }

  public void onBreaking() {
    MakeSqueakinSound();
  }

  private void MakeEngineSound() {
    makeASound(new Sound(/*...*/));
  }

  private void MakeSqueakinSound() {
    makeASound(new Sound(/*...*/));
  }
}
```

I've also described real life scenario in this [SO question](https://stackoverflow.com/questions/53797046/provide-extension-method-for-types-derived-from-this-type/53797582?noredirect=1#comment94446405_53797582)

**Comment by yaakov-h** [comment]
I don't think this is possible at the CLR level.

**Comment by brabebhin** [comment]
What if the extension method is defined in a different assembly?

**Comment by YairHalberstadt** [comment]
An extension method makes very little sense if it can only be used by members inheriting from the base class. As they correctly point out in the SO answers, the correct solution is to add another level of inheritance. That is exactly what inheritance is there for.

**Comment by yaakov-h** [comment]
I feel like what you really want to solve your problem would be mixins (#1014 and similar).

**Comment by Audionysos** [comment]
I think there are lot of proposals that could nicely solve the problem but they are much more complicated to implement. I wanted to make it simple so this one I believe
- Will not introduce any new syntax or any extra types (it's a matter of analyzing the context). 
- Will not collide with any existing feature.
- Can be resolved at compile time because the extension method could be used only in derived classes so CLR don't have to be involved here...

> What if the extension method is defined in a different assembly?

Anything else will work as usual. The code could be copied from .dll as well

> An extension method makes very little sense if it can only be used by members inheriting from the base class. As they correctly point out in the SO answers, the correct solution is to add another level of inheritance. That is exactly what inheritance is there for.

- The class could always explicitly expose used extension methods for descendants or even publicly - it's up to programmer what (s)he want to do with an extension.
- Deep inheritance is not good and does not allow such flexibility as this feature could do.
  - How could I use both methods of `MyCarSoundsExtensions` and `MyFriensCarSoundsExtensions` if they both would extend the `Car` class? Protected extension methods will allow you to use many extension classes from many different assemblies.
  - Often you wold extend a class just for single functionality while it provide many more that are not needed for you. With protected extension method you can provide tons of method and a class developer could chose only those that (s)he wants.

**Comment by TheUnlocked** [comment]
The idea that code could appear in your classes without you putting it there is a bit weird. With partial classes, C# requires that you specify when a class will be broken up, and even then it isn't typically encouraged to use partial classes when not using auto-generated code.

**Comment by jnm2** [comment]
@TheUnlocked Code appears in your class when adding a base class, but it does not appear in your class with DIM as far as I understand. In other words, you have to cast an instance of your class to the interface type in order to see the new members; you can't see them by dotting off your class type.

**Comment by Audionysos** [comment]
> The idea that code could appear in your classes without you putting it there is a bit weird. With partial classes, C# requires that you specify when a class will be broken up, and even then it isn't typically encouraged to use partial classes when not using auto-generated code.

Am I really that bad at expressing my thoughts? I'm not saying it should physically appear in your source file! That would be insane... You already have lot of language constructs that generate code for you like `int xyz { get; set; }` will generate two methods for you but you don't need to see it in a source file.

**Comment by TheUnlocked** [comment]
> I'm not saying it should physically appear in your source file! That would be insane... You already have lot of language constructs that generate code for you like `int xyz { get; set; }` will generate two methods for you but you don't need to see it in a source file.

With properties you do explicitly state that you want the getter and/or setter to be created. That's the `{ get; set; }` part. However, that's sort of besides the point.

In the example you provide (and that's all I can really base my opinions off of), the methods `MakeEngineSound` and `MakeSqueakinSound` magically appear within the class as a result of being defined in the extension. No intent is given to say that the functions should appear in the class, they just do because they're defined as extension methods. That's not really okay though; we can't just have functions, even private ones, appearing in classes out of nowhere without any demonstrated intent by the class developer for them to. Not only would it likely incur performance costs, but it could also potentially result in some kind of code injection through poorly-written reflection.

If you consider using one of the protected extension methods in the class to be demonstrating intent, that also seems a bit weird to me. It's highly unexpected for methods to pop in and out of existence depending only on whether or not you use them.

(To build, slightly tangentially, on that last point, I actually have similar concerns about how static code only runs if a class is referenced by something else that will also run, but changing that behavior at this point could be potentially breaking in weird edge cases with benefit in only very limited circumstances)

Edit: To clarify, when you extend a class, it is obviously assumed that you'll want everything from that class in your own. However, it is not necessarily assumed that you want all of the extra baggage from the protected extension methods in your class.

**Comment by Audionysos** [comment]
> If you consider using one of the protected extension methods in the class to be demonstrating intent

Yes. I actually wrote earlier "With protected extension method you can provide tons of method and a class developer could chose only those that (s)he wants."

> It's highly unexpected for methods to pop in and out of existence depending only on whether or not you use them.

Well, the expectations are quite subjective. I believe that someone who writes his first program in life, don't have any specific expectations at all from a language... The same applies for any new concept that you are not familiar with yet.
As another example I can tell you that when I first started using C# I was surprised that I have to call base constructor before doing anything else in my current constructor. I expected that I will be able to call in at the end of my new constructor but instead I was forced to change my strategy when translating a library form another language.
And finally, what is wrong with not having something that is not used? 

> I actually have similar concerns about how static code only runs if a class is referenced by something else that will also run, but changing that behavior at this point could be potentially breaking in weird edge cases with benefit in only very limited circumstances

I don't really get that. You can say the same about anything within the assembly, that if you don't use something, it's not used. What's special about static code in that matter? 

> To clarify, when you extend a class, it is obviously assumed that you'll want everything from that class in your own.

The problem is that this assumption is essentially wrong. There are millions of proves for this. 

The best example is probably WPF where you have tons of useful members on controls like `ScrollBar.FontSize`. This is not only redundant, but worse, it's misleading for the user of the API. 
And instead of focusing on things that are actually relevant to scrollbar itself I have to dig through an ocean of garbage. 

> However, it is not necessarily assumed that you want all of the extra baggage from the protected extension methods in your class.

I'm not sure about what extra baggage are you talking about? The hole point of my idea was to not introduce any unnecessary stuff like you are likely to do by inheriting from a class. My idea was to copy only those methods that are used inside your current the class.

I if you will use it or not would be up to you. I'm not talking about replacing the concept of inheritance but to supplement it. While inheritance is a core concept, it's not ideal and cannot solve all the problems, especially in it's current form because the world is not build this way but we are trying reflect the world in our programs.

**Comment by Thaina** [comment]
I support this idea. From the start we made extension method so that we don't need to create a whole class just for little helper function

Having anything defined in a different assembly should not be a problem at all. Derived class also able to access protected member of different assembly anyway. If that class can derived then it should be able to extended. If it `protected` not `private` it did means it just shouldn't access publicly. But extension method already is like a member of that class 

And we could call even private member by reflection. So there is no point to make a friction on extension method

**Comment by maazl** [comment]
### Better solution

I don not think the the purposed syntax for the extension is correct. **It makes no sense to expose an extension method as public while it is protected.** This raises several inconsistencies. First of all, the public method must be compilable although it calls a non-public method.

**A protected extension must itself be in a protected context,** i.e.:

```cs
public class Car {
  protected void makeASound(Sound s) {
    //...
  }
}

// My library
public class ModelA : Car {

  private void onRun() {
    this.MakeEngineSound();
    //...
  }

  protected static class MyCarSoundsExtensions {
    public static void MakeEngineSound(this Car c) {
      c.makeASound(new Sound(/*...*/));
    }
  }
}
```

The only change needed for this to work is that extension method lookup need to consider *nested statis classes* as well.
This would be helpful in other use cases too, e.g. where a class needs an extension only locally and the extension class could be nested and private.

The protected extensions can only be used from derived classes anyway so it is no restriction that the extensions can only be implemented in sub classes (or the class itself, of course).

Also note the `this.` qualifier. Extension methods can never be called without an explicit object. This should not be changed.

**Reply by Audionysos** [reply]
> It makes no sense to expose an extension method as public while it is protected.

I've used term "protected extension method" just for the title. My idea was not that this extension method will be a protected member - it should be defined as any other extension method i.e. should be public.  Sorry for confusion.

Again. The only difference from standard extension methods will be that you would use protected member of 'this' object inside the method. The compiler would detect use of protected members and mark this method as "protected" (in sense it uses protected members).
The compiler will allow you to use such method only internally - in a class that extends class of 'this' object used in the method, or in other words, in a context that has access to all the members used within the extension method:

    public class ModelA : Car {
    	private void onRun() {
    		this.MakeEngineSound(); //all good
    	}	
    }
    
    public class Chicken {
    	private void peckThePaint(Car c) {
    		c.MakeEngineSound(); //compiler error - this "protected" extension method may be used only in classes that derive from Car
    	}
    }
 
I'm not sure about your example. If you put this method in nested protected class it means only `ModelA` can use it, which defeats the whole idea. If you're ok with putting the method in there, why you would need extension method in the first place and not put it directly? 

    public class ModelA : Car {
    	protected void MakeEngineSound() {
    		this.makeASound(new Sound(/*...*/));
    	}
    
    	private void onRun() {
    		this.MakeEngineSound();
    	}   		
    }

The idea is that the `Car` is a closed code, but intended for inheritance and we will be making different models with direct inheritance like `ModelA : Car` or `ModelB : Car` and these models may use the same extension methods which are not defined either in `Car`, `ModelA` or `ModelB`. Otherwise we could just make `MyCar : Car` then `ModelA : MyCar` and `ModelB : MyCar`, but the problem with that is we are limiting ourselves only to what is defined in the `MyCar` class.
When someone makes `TheirCar : Car` with a single method that we think is better than the one we use in `MyCar`, we can't just simply borrow that single method to our class.
